### PR TITLE
Improved speed of frobenous norm for non-complex dtype

### PR DESCRIPTION
**Pull request by dylanbespalko**
In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for CUDA complex numbers is here: [pytorch-cuda-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cuda-strided-complex)

Changes:
[x] Fixed performance issue raise in #30704 so that non-complex numbers do not call `conj()` and `real()`.
[x] Fixed tensor_to_numpy() conversion likely broken by a `checkBackend()` in #27064.
[x] Fixed some ReduceOps and TensorCompare Ops that recently added a `checkBackend()`.
    - `checkBackend()` is replaced with a device type check and a layout check.
    - This ensures the ComplexCPU Type ID is supported. 
[x] Added AVX support for complex `exp()`, as requested in #755

### CODE REVIEW

**Comment by rgommers** [comment]
> Fixed tensor_to_numpy() conversion likely broken by a `checkBackend()` in #27064.

I think you meant gh-30281 here? Cc @ngoldbaum for this change

**Comment by ngoldbaum** [comment]
FWIW, here’s the deeplink to the relevant change from #30281: https://github.com/pytorch/pytorch/pull/30281/files#diff-d9318428a98d4f4d86210a8f4a736ebd.

Is the output of `tensor.type().backend()` not always the same as `tensor.options().backend()`? That would indicate deeper issues wrt the changes I made in #30281.

**Comment by ezyang** [comment]
@dylanbespalko Heads up that I will be out of country for three weeks so you'll need to find someone else to merge this PR.

File: torch/csrc/utils/tensor_numpy.cpp [hunk 1]
```diff
@@ -71,18 +71,15 @@ static std::vector<int64_t> seq_to_aten_shape(PyObject *py_seq) {
...
 
 PyObject* tensor_to_numpy(const at::Tensor& tensor) {
-  if (tensor.is_cuda()) {
-    throw TypeError(
-        "can't convert CUDA tensor to numpy. Use Tensor.cpu() to "
-        "copy the tensor to host memory first.");
+  if (tensor.device().type() != DeviceType::CPU) {
+      throw TypeError(
+        "can't convert %s device type tensor to numpy. Use Tensor.cpu() to "
+        "copy the tensor to host memory first.", tensor.device().type());
```
Comment (ezyang) [inline torch/csrc/utils/tensor_numpy.cpp]: Wow, does this really do printf style formatting strings? News to me.

**Review by ezyang — APPROVED** [review]

**Comment by dylanbespalko** [comment]
> Is the output of `tensor.type().backend()` not always the same as `tensor.options().backend()`? That would indicate deeper issues wrt the changes I made in #30281.

My limited understanding is that `Backend` dispatch was replaced with a `DeviceType` X `Layout` X `Dtype` multi-layered kernel dispatch.. I think `checkBackend(Backend::CPU)` corresponds to DeviceType::CPU, Layout::Strided, DType:(All internally supported dtypes), but I could be wrong.

Anyways, the backend checks are supposed to migrate into DeviceType, Layout, Dtype checks as described in [PyTorch Internals](http://blog.ezyang.com/2019/05/pytorch-internals/).

**Comment by dylanbespalko** [comment]
@ezyang,

Hopefully you are going on a much needed vacation. Happy holidays.

**Comment by dylanbespalko** [comment]
@VitalyFedyunin, @cpuhrsch, @gchanan,

Hey Guys, @ezyang is away on vacation. Can you forward me to someone that can merge this PR?

Pulling changes from master and re-running CI.

cc @ MuawizChaudhary

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit c803b06:


* **1/1** broken upstream at merge base 7a12ccd (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=7a12ccd0034198e45797b428b6cf86bda680325c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c803b06b63a3464749f2e0f0c389d92896706f26).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=78436641)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by dylanbespalko** [comment]
@ezyang,

I'm back from my FPGA adventure. I am looking to split my time between the CPU and FPGA for the next while and I plan to improve my CUDA coverage when I get a new PC. 

Let me know if this PR is too old.

**Comment by ezyang** [comment]
Nope, this pr is good, just gotta land it. Are you gonna blog about your FPGA adventures? :)

**Comment by dylanbespalko** [comment]
@ezyang,

Yes, I think I can write a good blog. I'll let you know when I post it.

**Pull request closed without merging**